src/multiexp.rs

```diff
@@ -193,107 +193,103 @@ impl DensityTracker {
+                    let exp = exp.as_ref()[0] % (1 << c);
+
+                    if exp != 0 {
+                        bases.add_assign_mixed(&mut buckets[(exp - 1) as usize])?;
                     } else {
-                        let mut exp = exp;
-                        exp.shr(skip);
-                        let exp = exp.as_ref()[0] % (1 << c);
-
-                        if exp != 0 {
-                            bases.add_assign_mixed(&mut buckets[(exp - 1) as usize])?;
-                        } else {
-                            bases.skip(1)?;
-                        }
+                        bases.skip(1)?;
                     }
                 }
             }
+        }
 
-            // Summation by parts
-            // e.g. 3a + 2b + 1c = a +
-            //                    (a) + b +
-            //                    ((a) + b) + c
-            let mut running_sum = G::Projective::zero();
-            for exp in buckets.into_iter().rev() {
-                running_sum.add_assign(&exp);
-                acc.add_assign(&running_sum);
-            }
+        // Summation by parts
+        // e.g. 3a + 2b + 1c = a +
+        //                    (a) + b +
+        //                    ((a) + b) + c
+        let mut running_sum = G::Projective::zero();
+        for exp in buckets.into_iter().rev() {
+            running_sum.add_assign(&exp);
+            acc.add_assign(&running_sum);
+        }
 
-            Ok(acc)
-        })
+        Ok(acc)
     };
 
-    skip += c;
-
-    if skip >= <G::Engine as ScalarEngine>::Fr::NUM_BITS {
-        // There isn't another region.
-        Box::new(this)
-    } else {
-        // There's another region more significant. Calculate and join it with
-        // this region recursively.
-        Box::new(
-            this.join(multiexp_inner(
-                pool,
-                bases,
-                density_map,
-                exponents,
+    let parts = (0..<G::Engine as ScalarEngine>::Fr::NUM_BITS)
+        .into_par_iter()
+        .step_by(c as usize)
+        .enumerate()
+        .map(|(i, skip)| {
+            let handle_trivial = if i == 0 { handle_trivial } else { false };
+            this(
+                bases.clone(),
+                density_map.clone(),
+                exponents.clone(),
                 skip,
-                c,
-                false,
-            ))
-            .map(move |(this, mut higher)| {
-                for _ in 0..c {
-                    higher.double();
-                }
+                handle_trivial,
+            )
+        })
+        .collect::<Vec<Result<_, _>>>();
 
-                higher.add_assign(&this);
+    let mut acc = <G as CurveAffine>::Projective::zero();
 
-                higher
-            }),
-        )
+    for (i, part) in parts.into_iter().enumerate().rev() {
+        acc.add_assign(&part?);
+        if i > 0 {
+            for _ in 0..c {
+                acc.double();
+            }
+        }
```

Comment:
I think this is right, but that it can maybe be made cleaner/clearer. Maybe by using the first element of the reversed parts the initial accumulator value, drop the conditional, and move the assignment after the doubling. Not blocking, but the more legible we can make this as an expression of intention, the better.

Reply:
I would still need the conditional, to avoid the doubeling on the last step though